### PR TITLE
Use buffer size at memory sync policy

### PR DESCRIPTION
**Pull request by bigsheeper**
issues: https://github.com/milvus-io/milvus/issues/22785

### CODE REVIEW

**Comment by bigsheeper** [comment]
rerun ut

File: internal/datanode/flow_graph_manager.go [hunk 1]
```diff
@@ -18,22 +18,92 @@ package datanode
...
+			value.(*dataSyncService).channel.getSyncFlag().syncKnown.Store(false)
+			return true
+		})
+		return
+	}
+
+	sort.Slice(channels, func(i, j int) bool {
+		return channels[i].bufferSize > channels[j].bufferSize
+	})
+	targetNum := math.Max(float64(len(channels))*0.5, 1) // sync half channels
```
Comment (xiaofan-luan) [inline internal/datanode/flow_graph_manager.go:101]: each time only set one channel to make it safer.
log print here.

File: internal/datanode/segment_sync_policy.go [hunk 2]
```diff
@@ -51,25 +53,25 @@ func syncPeriodically() segmentSyncPolicy {
...
+			return segments[i].memorySize > segments[j].memorySize
+		})
+		syncSegments := make([]UniqueID, 0)
+		syncNum := math.Max(float64(len(segments))*Params.DataNodeCfg.MemoryForceSyncSegmentRatio, 1.0)
+		for i := 0; i < int(syncNum); i++ {
+			if segments[i].memorySize < minSyncSize { // prevent generating too many small binlog
+				break
 			}
+			syncSegments = append(syncSegments, segments[i].segmentID)
 			log.Debug("sync segment due to memory usage is too high",
```
Comment (xiaofan-luan) [inline internal/datanode/segment_sync_policy.go:71]: use log.info

File: internal/util/paramtable/component_param.go [hunk 3]
```diff
@@ -1567,8 +1569,8 @@ type dataNodeConfig struct {
 
 	// memory management
 	MemoryForceSyncEnable       bool
-	MemoryForceSyncThreshold    float64
 	MemoryForceSyncSegmentRatio float64
```
Comment (xiaofan-luan) [inline internal/util/paramtable/component_param.go:1572]: I think number make more sense to me. by default could me only 1 segment

File: internal/datanode/channel_meta.go [hunk 4]
```diff
@@ -48,6 +50,12 @@ var (
 	newVarCharPrimaryKey = storage.NewVarCharPrimaryKey
 )
 
+type syncFlag struct {
+	needToSync *atomic.Bool
+	// syncKnown indicates sync is being executed, to prevent frequently syncing
+	syncKnown *atomic.Bool
```
Comment (xiaofan-luan) [inline internal/datanode/channel_meta.go:56]: syncKnown may not be necessary?
we can check all needToSync flag and see if there is anotjer sync?

Comment (bigsheeper) [reply internal/datanode/channel_meta.go:56]: Without syncKnown, sync policy would sync frequenly (per 200ms)?

File: internal/datanode/flow_graph_insert_buffer_node.go [hunk 5]
```diff
@@ -502,6 +506,7 @@ func (ibNode *insertBufferNode) Sync(fgMsg *flowGraphMsg, seg2Upload []UniqueID,
 		ibNode.channel.rollInsertBuffer(task.segmentID)
 		ibNode.channel.RollPKstats(task.segmentID, pkStats)
 		ibNode.channel.setSegmentLastSyncTs(task.segmentID, endPosition.GetTimestamp())
+		ibNode.channel.(*ChannelMeta).needToSync.Store(false)
```
Comment (czs007) [inline internal/datanode/flow_graph_insert_buffer_node.go:509]: better to move to head of this function ?

Comment (bigsheeper) [reply internal/datanode/flow_graph_insert_buffer_node.go]: done

**Pull request merged**